### PR TITLE
Provide better warnings when grouping multi clause functions

### PR DESCRIPTION
**Pull request by zacck**
##  Chore 
Relates #7512 
Providing a separate warning for cases where just the function name matches rather than the default warning. 

##  Test Plan 
Added a new unit test to cover the new  case

## Notes
I left the new arity of the function in the arguments in case we need to use it in the warning to the user

### CODE REVIEW

File: lib/elixir/src/elixir_def.erl [hunk 1]
```diff
@@ -396,6 +400,10 @@ format_error({clauses_with_defaults, {Kind, Name, Arity}}) ->
     "\n"
    "~ts ~ts/~B has multiple clauses and defines defaults in one or more clauses", [Kind, Name, Arity]);
 
+format_error({grouped_clause, {Kind, Name, _OtherArity, OrigLine, OrigFile}}) ->
+   io_lib:format("clauses with the same name should be grouped together, ~ts ~ts was previously defined (~ts:~B)",
```
Comment (whatyouhide) [inline lib/elixir/src/elixir_def.erl]: I think we should wrap the `def foo` part in quotes (it reads a bit confusing in the error message in the test below). Maybe we can include the arity too?

> "def foo/2" was previously defined

Thoughts?

Comment (zacck) [reply lib/elixir/src/elixir_def.erl]: I think the arity will help in identifying which clause needs to be corrected and I  agree with you on the quotes let me put in some time and I'll push those fixes

Comment (josevalim) [reply lib/elixir/src/elixir_def.erl]: I think we can improve both warnings.

Let's call the first `ungrouped_arity` and the other one `ungrouped_name`. For the arity, we should say:

> clauses with the same name and arity (number of arguments) should be grouped together, \"~ts ~ts/~B\" was previously defined (~ts:~B)

For `ungrouped_name`, we will say:

> clauses with the same name should be grouped together, \"~ts ~ts/~B\" was previously defined (~ts:~B)

File: lib/elixir/src/elixir_def.erl [hunk 2]
```diff
@@ -301,8 +301,12 @@ check_valid_kind(Meta, File, Name, Arity, Kind, StoredKind) ->
 
 check_valid_clause(Meta, File, Name, Arity, Kind, Set, StoredMeta, StoredFile) ->
   case ets:lookup_element(Set, ?last_def, 2) of
-    {Name, Arity} -> ok;
     none -> ok;
+    {Name, Arity} -> ok;
+    {Name, OtherArity} ->
+      Relative = elixir_utils:relative_to_cwd(StoredFile),
+      elixir_errors:form_warn(Meta, File, ?MODULE,
+        {grouped_clause, {Kind, Name, OtherArity, ?line(StoredMeta), Relative}});
```
Comment (josevalim) [inline lib/elixir/src/elixir_def.erl]: We should pass `Arity` here and not `OtherArity` and make use of `Arity` in the error message.

Comment (zacck) [reply lib/elixir/src/elixir_def.erl]: Ah

File: lib/elixir/src/elixir_def.erl (hunk 1 above)
Comment (zacck) [reply lib/elixir/src/elixir_def.erl]: @josevalim  woop woop! yours looks way cooler I like this one

**Comment by michalmuskala** [comment]
Thank you! :heart:

**Comment by zacck** [comment]
Its been a pleasure, hunting the next issue!

**Pull request merged**